File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/config/MybatisProperties.java
package top.bettercode.summer.data.jpa.config;

import java.util.Properties;
import org.apache.ibatis.session.Configuration;
import org.apache.ibatis.type.TypeHandler;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.NestedConfigurationProperty;

/**
 * Configuration properties for MyBatis.
 *
 * @author Eddú Meléndez
 * @author Kazuki Shimizu
 */
@ConfigurationProperties(prefix = MybatisProperties.MYBATIS_PREFIX)
public class MybatisProperties {

  public static final String MYBATIS_PREFIX = "spring.data.jpa.mybatis";

  /**
   * Location of MyBatis xml config file.
   */
  private String configLocation;

  /**
   * Locations of MyBatis mapper files.
   */
  private String[] mapperLocations;

  /**
   * Packages to search type aliases. (Package delimiters are ",; \t\n")
   */
  private String typeAliasesPackage;

  private Class<?>[] typeAliases;

  /**
   * The super class for filtering type alias. If this not specifies, the MyBatis deal as type alias
   * all classes that searched from typeAliasesPackage.
   */
  private Class<?> typeAliasesSuperType;


  /**
   * Packages to search for type handlers. (Package delimiters are ",; \t\n")
   */
  private String typeHandlersPackage;

  private Class<TypeHandler<?>>[] typeHandlerClasses;

  /**
   * Indicates whether perform presence check of the MyBatis xml config file.
   */
  private boolean checkConfigLocation = false;

  /**
   * Externalized properties for MyBatis configuration.
   */
  private Properties configurationProperties;

  /**
   * 结果集转换是否使用 TupleTransformer
   */
  private Boolean useTupleTransformer = false;
  /**
   * A Configuration object for customize default settings. If {@link #configLocation} is specified,
   * this property is not used.
   */
  @NestedConfigurationProperty
  private Configuration configuration;

  public String getConfigLocation() {
    return this.configLocation;
  }

  public void setConfigLocation(String configLocation) {
    this.configLocation = configLocation;
  }

  public String[] getMapperLocations() {
    return this.mapperLocations;
  }

  public void setMapperLocations(String[] mapperLocations) {
    this.mapperLocations = mapperLocations;
  }

  public String getTypeHandlersPackage() {
    return this.typeHandlersPackage;
  }

  public void setTypeHandlersPackage(String typeHandlersPackage) {
    this.typeHandlersPackage = typeHandlersPackage;
  }

  public Class<TypeHandler<?>>[] getTypeHandlerClasses() {
    return typeHandlerClasses;
  }

  public MybatisProperties setTypeHandlerClasses(
      Class<TypeHandler<?>>[] typeHandlerClasses) {
    this.typeHandlerClasses = typeHandlerClasses;
    return this;
  }

  public String getTypeAliasesPackage() {
    return this.typeAliasesPackage;
  }

  public void setTypeAliasesPackage(String typeAliasesPackage) {
    this.typeAliasesPackage = typeAliasesPackage;
  }

  public Class<?>[] getTypeAliases() {
    return typeAliases;
  }

  public void setTypeAliases(Class<?>[] typeAliases) {
    this.typeAliases = typeAliases;
  }

  public Class<?> getTypeAliasesSuperType() {
    return typeAliasesSuperType;
  }

  public void setTypeAliasesSuperType(Class<?> typeAliasesSuperType) {
    this.typeAliasesSuperType = typeAliasesSuperType;
  }

  public boolean isCheckConfigLocation() {
    return this.checkConfigLocation;
  }

  public void setCheckConfigLocation(boolean checkConfigLocation) {
    this.checkConfigLocation = checkConfigLocation;
  }

  public Properties getConfigurationProperties() {
    return configurationProperties;
  }

  public void setConfigurationProperties(Properties configurationProperties) {
    this.configurationProperties = configurationProperties;
  }

  public Boolean getUseTupleTransformer() {
    return useTupleTransformer;
  }

  public void setUseTupleTransformer(Boolean useTupleTransformer) {
    this.useTupleTransformer = useTupleTransformer;
  }

  public Configuration getConfiguration() {
    return configuration;
  }

  public void setConfiguration(Configuration configuration) {
    this.configuration = configuration;
  }

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/config/MybatisProperties.kt
package top.bettercode.summer.data.jpa.config

import org.apache.ibatis.session.Configuration
import org.apache.ibatis.type.TypeHandler
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.NestedConfigurationProperty
import java.util.*

/**
 * Configuration properties for MyBatis.
 *
 * @author Eddú Meléndez
 * @author Kazuki Shimizu
 */
@ConfigurationProperties(prefix = MybatisProperties.MYBATIS_PREFIX)
open class MybatisProperties {
    /**
     * Location of MyBatis xml config file.
     */
    var configLocation: String? = null

    /**
     * Locations of MyBatis mapper files.
     */
    var mapperLocations: Array<String> = arrayOf()

    /**
     * Packages to search type aliases. (Package delimiters are ",; \t\n")
     */
    var typeAliasesPackage: String? = null
    var typeAliases: Array<Class<*>> = arrayOf()

    /**
     * The super class for filtering type alias. If this not specifies, the MyBatis deal as type alias
     * all classes that searched from typeAliasesPackage.
     */
    var typeAliasesSuperType: Class<*>? = null

    /**
     * Packages to search for type handlers. (Package delimiters are ",; \t\n")
     */
    var typeHandlersPackage: String? = null
    var typeHandlerClasses: Array<Class<TypeHandler<*>>> = arrayOf()


    /**
     * Externalized properties for MyBatis configuration.
     */
    var configurationProperties: Properties? = null

    /**
     * 结果集转换是否使用 TupleTransformer,true时存在类型转换不兼容问题，如日期时间格式化
     */
    var useTupleTransformer = false

    /**
     * A Configuration object for customize default settings. If [.configLocation] is specified,
     * this property is not used.
     */
    @NestedConfigurationProperty
    var configuration: Configuration? = null

    companion object {
        const val MYBATIS_PREFIX = "summer.data.jpa.mybatis"
    }
}
